server/db.go

```diff
@@ -419,49 +643,198 @@ func (c *Server) rotateIssuers() (err error) {
+		issuer.Version,
+		issuer.ExpiresAt,
+		issuer.Buffer,
+		issuer.Duration,
+	)
+
+	// get the newly inserted issuer identifier
+	if err := row.Scan(&issuer.ID); err != nil {
+		tx.Rollback()
+		return fmt.Errorf("failed to get v3 issuer id: %w", err)
 	}
 
-	signingKeyTxt, err := signingKey.MarshalText()
-	if err != nil {
-		c.Logger.Error("Error marshalling signing key")
-		return err
+	if err := txPopulateIssuerKeys(c.Logger, tx, issuer); err != nil {
+		tx.Rollback()
+		return fmt.Errorf("failed to close rows on v3 issuer creation: %w", err)
 	}
+	queryTimer.ObserveDuration()
+	return tx.Commit()
+}
 
-	queryTimer := prometheus.NewTimer(createIssuerDBDuration)
-	rows, err := c.db.Query(
-		`INSERT INTO issuers(issuer_type, issuer_cohort, signing_key, max_tokens, expires_at, version) VALUES ($1, $2, $3, $4, $5, 2)`,
-		issuerType,
-		issuerCohort,
-		signingKeyTxt,
-		maxTokens,
-		expiresAt,
+// on the transaction, populate v3 issuer keys for the v3 issuer
+func txPopulateIssuerKeys(logger *logrus.Logger, tx *sqlx.Tx, issuer Issuer) error {
+	var (
+		duration *timeutils.ISODuration
+		err      error
 	)
+
+	if issuer.Version == 3 {
+		// get the duration from the issuer
+		duration, err = timeutils.ParseDuration(issuer.Duration)
+		if err != nil {
+			return fmt.Errorf("failed to parse issuer duration: %w", err)
+		}
+	}
+
+	// v1/v2 issuers only have a buffer of 1
+	if issuer.Version < 3 {
+		issuer.Buffer = 1
+	}
+
+	start := issuer.ValidFrom
+	i := 0
+	// time to create the keys associated with the issuer
+	if issuer.Keys == nil || len(issuer.Keys) == 0 {
+		issuer.Keys = []IssuerKeys{}
+	} else {
+		// if the issuer has keys already, start needs to be the last item in slice
+		start = issuer.Keys[len(issuer.Keys)-1].EndAt
+		i = len(issuer.Keys)
+	}
+
+	valueFmtStr := ""
+
+	var keys = []issuerKeys{}
+
+	// for i in buffer, create signing keys for each
+	for ; i < issuer.Buffer; i++ {
+		end := new(time.Time)
+		if duration != nil {
+			// start/end, increment every iteration
+			end, err = duration.From(*start)
+			if err != nil {
+				tx.Rollback()
+				return fmt.Errorf("unable to calculate end time: %w", err)
+			}
+		}
+
+		signingKey, err := crypto.RandomSigningKey()
+		if err != nil {
+			logger.Error("Error generating key")
+			tx.Rollback()
+			return err
+		}
+
+		signingKeyTxt, err := signingKey.MarshalText()
+		if err != nil {
+			logger.Error("Error marshalling signing key")
+			tx.Rollback()
+			return err
+		}
+
+		pubKeyTxt, err := signingKey.PublicKey().MarshalText()
+		if err != nil {
+			logger.Error("Error marshalling public key")
+			tx.Rollback()
+			return err
+		}
+
+		keys = append(keys, issuerKeys{
+			SigningKey: signingKeyTxt,
+			PublicKey:  string(pubKeyTxt),
+			Cohort:     issuer.IssuerCohort,
+			IssuerID:   issuer.ID,
+			StartAt:    start,
+			EndAt:      end,
+		})
+
+		if start != nil && !(*start).Equal(*issuer.ValidFrom) {
+			valueFmtStr += ", "
+		}
+		valueFmtStr += "($1, $2, $3, $4, $5, $6)"
```

Comment:
Do these params need to increment when using multiple insert i.e. final statement  for two rows would be  "($1, $2, $3, $4, $5, $6), ($7, $8, $9, $10, $11, $12)"

Reply:
https://github.com/brave-intl/challenge-bypass-server/pull/145/commits/be6d98acc75b686f32c473183ee19ff99a79e225